src/Storages/MergeTree/MergeTreeReaderStream.cpp

```diff
@@ -138,6 +138,27 @@ std::pair<size_t, size_t> MergeTreeReaderStream::getRightOffsetAndBytesRange(siz
     else
     {
         right_offset = marks_loader.getMark(result_right_mark).offset_in_compressed_file;
+        /// In LowCardinality dictionary several consecutive marks can point to
+        /// the same offset. So to get true bytes offset we have to get first
+        /// non-equal mark.
+        /// Example:
+        ///  Mark 186, points to [2003111, 0]
+        ///  Mark 187, points to [2003111, 0]
+        ///  Mark 188, points to [2003111, 0]
+        ///  Mark 189, points to [2003111, 0]
+        ///  Mark 190, points to [2003111, 0]
+        ///  Mark 191, points to [2003111, 0]
+        ///  Mark 192, points to [2081424, 0]
+        ///  Mark 193, points to [2081424, 0]
+        ///  Mark 194, points to [2081424, 0]
+
+        while (result_right_mark != 0 && result_right_mark < marks_count
+               && marks_loader.getMark(result_right_mark - 1).offset_in_compressed_file == right_offset)
```

Comment:
Maybe it's possible to lower the number of `getMark` calls to avoid getting the same mark twice?

Reply:
Removed redundant loop. Now it's also possible to make it better forwarding `substream path` and check for right mark only in case of LowCardinality or non-zero offset in decompressed block.

Reply:
Maybe it's better to get rid of binary search and just go to the right until we get appropriate mark?